subprojects/core/src/main/java/org/gradle/api/internal/project/CrossProjectModelAccess.java

```diff
@@ -62,6 +62,13 @@ public interface CrossProjectModelAccess {
      */
     Set<? extends ProjectInternal> getAllprojects(ProjectInternal referrer, ProjectInternal relativeTo);
 
+    /**
+     * For usage only at the {@code Gradle} level, when all projects are requested before the root project evaluation.
+     *
+     * @return All projects decorated accordingly. <b>Each</b> project will be decorated regardless of the referring relation.
```

Comment:
❓  What is the reason for decorating the root project as well?
Is this different in behavior from `getAllProjects(rootProject, rootProject)` ?

Reply:
@alllex This code was changed by force push(sorry, did it before seen your comments), PTAL at new implemenation.

> What is the reason for decorating the root project as well?

Consider the test case:
https://github.com/gradle/gradle/pull/29917/files#diff-f7a18918781bafa6dc24a16986e375355e5929b16fb6a0cf99efe13b26b9985fR221

To see `lifecycle.beforeProject` result in `gradle.allprojects` for root project, root project has to be decorated